From 8f4a69dc7f0e9b4593cc54c501273783426b9838 Mon Sep 17 00:00:00 2001
From: q66 <q66@chimera-linux.org>
Date: Thu, 2 Mar 2023 02:00:22 +0100
Subject: [PATCH] whereis: skip if we can't query manpage paths

Also use man -w instead of manpath, as with mandoc manpath does
not exist.

Also do not fail if we can't query the paths, it may be just so
that e.g. base-man in chimera is not installed, and we still
want to print other results (i.e. pretend there are no manpaths).
---
 patches/src.freebsd.patch                 | 18 +++++++++++++++++-
 src.freebsd/miscutils/whereis/pathnames.h |  2 +-
 src.freebsd/miscutils/whereis/whereis.c   |  2 +-
 3 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/patches/src.freebsd.patch b/patches/src.freebsd.patch
index 0e6a53c..7214d7d 100644
--- a/patches/src.freebsd.patch
+++ b/patches/src.freebsd.patch
@@ -7696,7 +7696,7 @@
  #define PATH_SOURCES					\
  "/usr/src/bin:/usr/src/usr.bin:/usr/src/sbin:"		\
  "/usr/src/usr.sbin:/usr/src/libexec:"			\
-@@ -41,6 +42,9 @@
+@@ -41,12 +42,15 @@
  "/usr/src/secure/sbin:/usr/src/secure/usr.sbin:"	\
  "/usr/src/secure/libexec:/usr/src/crypto:"		\
  "/usr/src/games"
@@ -7706,6 +7706,13 @@
  
  /* Each subdirectory of PATH_PORTS will be appended to PATH_SOURCES. */
  #define PATH_PORTS "/usr/ports"
+ 
+ /* How to query the current manpath. */
+-#define MANPATHCMD "manpath -q"
++#define MANPATHCMD "command -v man > /dev/null && man -w || :"
+ 
+ /* How to obtain the location of manpages, and how to match this result. */
+ #define MANWHEREISCMD "man -S1:8:6 -w %s 2>/dev/null"
 --- src.orig/miscutils/whereis/whereis.1
 +++ src.freebsd/miscutils/whereis/whereis.1
 @@ -60,11 +60,7 @@
@@ -7798,6 +7805,15 @@
  		nele = 0;
  		decolonify(b, &bindirs, &nele);
  		bindirs = realloc(bindirs, (nele + 2) * sizeof(char *));
+@@ -296,7 +293,7 @@
+ 	if (!mandirs) {
+ 		if ((p = popen(MANPATHCMD, "r")) == NULL)
+ 			err(EX_OSERR, "cannot execute manpath command");
+-		if (fgets(buf, BUFSIZ - 1, p) == NULL ||
++		if ((fgets(buf, BUFSIZ - 1, p) == NULL && !feof(p)) ||
+ 		    pclose(p))
+ 			err(EX_OSERR, "error processing manpath results");
+ 		if ((b = strchr(buf, '\n')) != NULL)
 @@ -315,7 +312,7 @@
  			abort();
  		nele = 0;
diff --git a/src.freebsd/miscutils/whereis/pathnames.h b/src.freebsd/miscutils/whereis/pathnames.h
index 8b78738..d6c3995 100644
--- a/src.freebsd/miscutils/whereis/pathnames.h
+++ b/src.freebsd/miscutils/whereis/pathnames.h
@@ -50,7 +50,7 @@
 #define PATH_PORTS "/usr/ports"
 
 /* How to query the current manpath. */
-#define MANPATHCMD "manpath -q"
+#define MANPATHCMD "command -v man > /dev/null && man -w || :"
 
 /* How to obtain the location of manpages, and how to match this result. */
 #define MANWHEREISCMD "man -S1:8:6 -w %s 2>/dev/null"
diff --git a/src.freebsd/miscutils/whereis/whereis.c b/src.freebsd/miscutils/whereis/whereis.c
index 38a841d..b07a6f8 100644
--- a/src.freebsd/miscutils/whereis/whereis.c
+++ b/src.freebsd/miscutils/whereis/whereis.c
@@ -293,7 +293,7 @@ defaults(void)
 	if (!mandirs) {
 		if ((p = popen(MANPATHCMD, "r")) == NULL)
 			err(EX_OSERR, "cannot execute manpath command");
-		if (fgets(buf, BUFSIZ - 1, p) == NULL ||
+		if ((fgets(buf, BUFSIZ - 1, p) == NULL && !feof(p)) ||
 		    pclose(p))
 			err(EX_OSERR, "error processing manpath results");
 		if ((b = strchr(buf, '\n')) != NULL)
